### PR TITLE
Add support for block-level independent formatting contexts

### PR DESCRIPTION
**Pull request by SimonSapin**
… and various drive-by changes.

### CODE REVIEW

**Comment by SimonSapin** [comment]
@nox, with this `display: flow-root` creates a block that, unlike same-formatting-context blocks, should not collide with floats:

https://drafts.csswg.org/css2/visuren.html#floats
>  The border box of a table, a block-level replaced element, or an element in the normal flow that establishes a new block formatting context (such as an element with 'overflow' other than 'visible') must not overlap the margin box of any floats in the same block formatting context as the element itself.

File: victor/src/layout/flow/mod.rs [hunk 1]
```diff
@@ -139,19 +138,24 @@ impl BlockLevelBox {
...
                     absolutely_positioned_fragments,
                     style,
-                    contents,
-                )
-            }
-            BlockLevelBox::Independent { style: _, contents } => {
-                // FIXME
-                match *contents {}
-            }
+                    move |containing_block| contents.layout(containing_block),
```
Comment (nox) [inline victor/src/layout/flow/mod.rs:156]: I am really confused about this code, we don't pass `tree_rank` so what happens to abspos elements in independent formatting contexts?

Comment (SimonSapin) [reply victor/src/layout/flow/mod.rs:156]: `tree_rank` is used in `BlockContainer::layout` to mark abspos elements in this block container apart from those in siblings. `BlockFormattingContext::layout` does not take a `tree_rank` parameter and instead uses `let dummy_tree_rank = 0;`. When a block formatting context contains a block container it contains nothing else, so there is no sibling to distinguish from.

Comment (nox) [reply victor/src/layout/flow/mod.rs:156]: So let's say there is an IndependentFormattingContext somewhere in a BlockContainer, how is the tree rank of that independent formatting context propagated to its abspos descendants?

Comment (nox) [reply victor/src/layout/flow/mod.rs:156]: I feel like somewhere, we are lacking some tree rank propagation for [this code](https://github.com/SimonSapin/victor/blob/03888c56613bc305e8d40c571a08ed9d0a347d31/victor/src/layout/flow/mod.rs#L81-L87) to work properly.

Comment (nox) [reply victor/src/layout/flow/mod.rs:156]: We can discuss about that tomorrow.

Comment (SimonSapin) [reply victor/src/layout/flow/mod.rs:156]: I think you’re right, we’re missing code is some places both to count the rank and to adjust static positions to account for coordinate origin change.

Comment (nox) [reply victor/src/layout/flow/mod.rs:156]: I have a commit for it, which is nice.

**Pull request merged**